apps/els_lsp/src/els_bsp_provider.erl

```diff
@@ -0,0 +1,262 @@
+-module(els_bsp_provider).
+
+-behaviour(els_provider).
+
+%% API
+-export([ start/1
+        , maybe_start/1
+        , info/1
+        , request/2
+        ]).
+
+%% els_provider functions
+-export([ is_enabled/0
+        , init/0
+        , handle_request/2
+        , handle_info/2
+        ]).
+
+
+%%==============================================================================
+%% Includes
+%%==============================================================================
+-include("els_lsp.hrl").
+-include_lib("kernel/include/logger.hrl").
+
+%%==============================================================================
+%% Types
+%%==============================================================================
+-type state() :: #{ running          := boolean()         % BSP server running?
+                  , root_uri         := uri() | undefined % the root uri
+                  , pending_requests := list()            % pending requests
+                  , pending_sources  := list()            % pending source info
+                  }.
+-type method() :: binary().
+-type params() :: map() | null.
+-type request_response() :: {reply, any()} | {error, any()}.
+-type request_id() :: {reference(), reference(), atom() | pid()}.
+-type from() :: {pid(), reference()} | self.
+-type config() :: map().
+-type info_item() :: is_running.
+
+%%==============================================================================
+%% API
+%%==============================================================================
+-spec start(uri()) -> {ok, config()} | {error, term()}.
+start(RootUri) ->
+  els_provider:handle_request(?MODULE, {start, #{ root => RootUri }}).
+
+-spec maybe_start(uri()) -> {ok, config()} | {error, term()} | disabled.
+maybe_start(RootUri) ->
+  case els_config:get(bsp_enabled) of
+    false ->
+      disabled;
+    X when X =:= true orelse X =:= auto ->
+      start(RootUri)
+  end.
+
+-spec info(info_item()) -> any().
+info(Item) ->
+  case els_provider:handle_request(?MODULE, {info, #{ item => Item}}) of
+    {ok, Result} ->
+      Result;
+    {error, badarg} ->
+      erlang:error(badarg, [Item])
+  end.
+
+-spec request(method(), params()) -> request_response().
+request(Method, Params) ->
+  RequestId = send_request(Method, Params),
+  wait_response(RequestId, infinity).
+
+-spec send_request(method(), params()) -> request_id().
+send_request(Method, Params) ->
```

Comment:
My gut feeling is that we should be able to do this with higher-level OTP primitives instead of expicit monitoring and receive, but we can tackle this as an eventual follow up.

Reply:
We should probably turn it into a gen_server and use the same send_request/receive_request etc. functionality as the client already uses.